File: voice-agent-webrtc-realtime/frontend/lib/tools.ts
import { sendEvent, type RealtimeSession } from "@/lib/realtime";
import type { Mode, Order, OrderItem } from "@/lib/types";

/** A receipt line as shown in OrderReceipt, with persistence state. */
export interface ReceiptOrder {
  order_id: string | null; // null until the backend assigns one
  items: OrderItem[];
  status: "pending" | "persisted" | "closed" | "error";
}

export interface ToolContext {
  mode: Mode;
  /** Optimistically update the receipt before the backend responds. */
  onOptimistic: (items: OrderItem[]) => void;
  /** Backend assigned/persisted an order id. */
  onPersisted: (order: Order) => void;
  onClosed: (order: Order) => void;
  onError: (message: string) => void;
}

interface FunctionCallEvent {
  type: string;
  name: string;
  call_id: string;
  arguments: string; // JSON string
}

function isFunctionCallDone(event: Record<string, unknown>): boolean {
  return event.type === "response.function_call_arguments.done";
}

/**
 * Handle a Realtime data-channel event. If it is a function call, execute it
 * against the backend HTTP API and return the result to the model via a
 * function_call_output item + response.create.
 *
 * Returns true if the event was a (handled) function call.
 */
export async function handleToolEvent(
  session: RealtimeSession,
  event: Record<string, unknown>,
  ctx: ToolContext,
): Promise<boolean> {
  if (!isFunctionCallDone(event)) return false;
  const call = event as unknown as FunctionCallEvent;
  const args = JSON.parse(call.arguments) as Record<string, unknown>;

  let output: Record<string, unknown>;
  try {
    output = await dispatch(call.name, args, ctx);
  } catch (err) {
    ctx.onError(String(err));
    output = { ok: false, error: String(err) };
  }

  sendEvent(session, {
    type: "conversation.item.create",
    item: {
      type: "function_call_output",
      call_id: call.call_id,
      output: JSON.stringify(output),
    },
  });
  sendEvent(session, { type: "response.create" });
  return true;
}

async function dispatch(
  name: string,
  args: Record<string, unknown>,
  ctx: ToolContext,
): Promise<Record<string, unknown>> {
  if (name === "add_order") {
    const items = args.items as OrderItem[];
    ctx.onOptimistic(items); // instant UI update, no Sheets round-trip
    const res = await fetch("/api/orders", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        mode: ctx.mode,
        customer_label: (args.customer_label as string) ?? null,
        language: args.language,
        items,
      }),
    });
    if (!res.ok) throw new Error(`add_order failed: ${res.status}`);
    const order = (await res.json()) as Order;
    ctx.onPersisted(order);
    return { ok: true, order_id: order.order_id };
  }

  if (name === "update_order") {
    const res = await fetch(`/api/orders/${args.order_id}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        items: args.items,
        notes: (args.notes as string) ?? null,
      }),
    });
    if (!res.ok) throw new Error(`update_order failed: ${res.status}`);
    const order = (await res.json()) as Order;
    ctx.onPersisted(order);
    return { ok: true, order_id: order.order_id };
  }

  if (name === "close_order") {
    const res = await fetch(`/api/orders/${args.order_id}/close`, {
      method: "POST",
    });
    if (!res.ok) throw new Error(`close_order failed: ${res.status}`);
    const order = (await res.json()) as Order;
    ctx.onClosed(order);
    return { ok: true, order_id: order.order_id, status: "closed" };
  }

  if (name === "list_orders") {
    const limit = (args.limit as number) ?? 10;
    const res = await fetch(`/api/orders/recent?limit=${limit}`);
    if (!res.ok) throw new Error(`list_orders failed: ${res.status}`);
    return (await res.json()) as Record<string, unknown>;
  }

  throw new Error(`unknown tool: ${name}`);
}

File: voice-agent-conoha-l4/agent/app/tools.py
"""Tool definitions and dispatcher for the agent.

Tools are described in OpenAI-compatible JSON schema (the format vLLM
accepts via `--enable-auto-tool-choice`). Dispatch translates the LLM's
tool_call into an HTTP call against the backend.
"""
from typing import Any

import httpx

from app import settings

_ITEM_SCHEMA = {
    "type": "object",
    "properties": {
        "name": {"type": "string"},
        "qty": {"type": "integer", "minimum": 1},
        "note": {"type": ["string", "null"]},
    },
    "required": ["name", "qty"],
}

OPENAI_TOOLS: list[dict[str, Any]] = [
    {
        "type": "function",
        "function": {
            "name": "add_order",
            "description": "新しい注文を業務システムに追加する。",
            "parameters": {
                "type": "object",
                "properties": {
                    "items": {"type": "array", "items": _ITEM_SCHEMA},
                    "customer_label": {"type": "string"},
                    "language": {"type": "string", "enum": ["ja", "en", "ko"]},
                },
                "required": ["items", "language"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "update_order",
            "description": "直前の注文の数量や品目を変更する。items は変更後の全品目。",
            "parameters": {
                "type": "object",
                "properties": {
                    "order_id": {"type": "string"},
                    "items": {"type": "array", "items": _ITEM_SCHEMA},
                    "notes": {"type": ["string", "null"]},
                },
                "required": ["order_id", "items"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "close_order",
            "description": "注文を確定する。",
            "parameters": {
                "type": "object",
                "properties": {"order_id": {"type": "string"}},
                "required": ["order_id"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "list_orders",
            "description": "当日の最近の注文を確認する。",
            "parameters": {
                "type": "object",
                "properties": {"limit": {"type": "integer", "default": 10}},
                "required": [],
            },
        },
    },
]


class ToolExecutor:
    def __init__(self, mode: str, http_client: httpx.AsyncClient | None = None) -> None:
        self._mode = mode
        self._http = http_client or httpx.AsyncClient(timeout=8.0)

    @staticmethod
    def _extract_4xx(r) -> dict:
        """Parse a 4xx response body into a structured error result."""
        try:
            err = r.json()
        except Exception:
            err = {"detail": r.text}
        return {"ok": False, "error": "validation failed", "detail": err.get("detail")}

    async def dispatch(self, name: str, args: dict) -> dict:
        if name == "add_order":
            body = {
                "mode": self._mode,
                "language": args.get("language", "ja"),
                "customer_label": args.get("customer_label"),
                "items": args["items"],
            }
            r = await self._http.post(f"{settings.BACKEND_URL}/api/orders", json=body)
            if 400 <= r.status_code < 500:
                return self._extract_4xx(r)
            r.raise_for_status()
            order = r.json()
            return {"ok": True, "order_id": order["order_id"]}

        if name == "update_order":
            order_id = args["order_id"]
            body = {"items": args["items"], "notes": args.get("notes")}
            r = await self._http.patch(
                f"{settings.BACKEND_URL}/api/orders/{order_id}", json=body
            )
            if 400 <= r.status_code < 500:
                return self._extract_4xx(r)
            r.raise_for_status()
            return {"ok": True, "order_id": r.json()["order_id"]}

        if name == "close_order":
            order_id = args["order_id"]
            r = await self._http.post(
                f"{settings.BACKEND_URL}/api/orders/{order_id}/close"
            )
            if 400 <= r.status_code < 500:
                return self._extract_4xx(r)
            r.raise_for_status()
            return {"ok": True, "order_id": order_id, "status": "closed"}

        if name == "list_orders":
            limit = args.get("limit", 10)
            r = await self._http.get(
                f"{settings.BACKEND_URL}/api/orders/recent", params={"limit": limit}
            )
            r.raise_for_status()
            return r.json()

        raise ValueError(f"unknown tool: {name}")

    async def aclose(self) -> None:
        await self._http.aclose()
